chainerx_cc/chainerx/routines/math.cc

```diff
@@ -513,6 +513,53 @@ Array Sum(const Array& a, const OptionalAxes& axis, bool keepdims) {
     return out;
 }
 
+Array Prod(const Array& a, const OptionalAxes& axis, bool keepdims) {
+    Axes sorted_axis = internal::GetSortedAxesOrAll(axis, a.ndim());
+
+    // Decide the output dtype for integral input dtype.
+    Dtype out_dtype{};
+    switch (GetKind(a.dtype())) {
+        case DtypeKind::kBool:
+        case DtypeKind::kInt:  // fallthrough
+            out_dtype = Dtype::kInt64;
+            break;
+        case DtypeKind::kUInt:
+            out_dtype = Dtype::kInt64;  // TODO(niboshi): This should be kUInt64
+            break;
+        default:
+            out_dtype = a.dtype();
+    }
+
+    Array out = internal::EmptyReduced(a.shape(), out_dtype, sorted_axis, keepdims, a.device());
+    {
+        NoBackpropModeScope scope{};
+        a.device().backend().CallKernel<ProdKernel>(a, sorted_axis, out);
+    }
+
+    BackwardBuilder bb{"prod", a, out};
+    if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
+        bt.Define([sorted_axis, a_tok = bb.RetainInput(0), out_tok = bb.RetainOutput(0), in_shape = a.shape(), keepdims](
+                          BackwardContext& bctx) {
+            const Array& gout = *bctx.output_grad();
+            const Array& a = bctx.GetRetainedInput(a_tok);
+            const Array& out = bctx.GetRetainedOutput(out_tok);
+            CHAINERX_ASSERT(std::is_sorted(sorted_axis.begin(), sorted_axis.end()));
+
+            if (!(in_shape.ndim() == 0 || sorted_axis.empty() || keepdims)) {
+                Shape out_shape_broadcastable = gout.shape();
+                for (auto axis : sorted_axis) {
+                    out_shape_broadcastable.insert(out_shape_broadcastable.begin() + axis, 1);
+                }
+                bctx.input_grad() = (out.Reshape(out_shape_broadcastable) * gout.Reshape(out_shape_broadcastable) / a);
+            } else {
+                bctx.input_grad() = out * gout / a;
```

Comment:
Could you avoid `0 / 0`?  There are at least two ways (#3031, #5094) to implement backward that allows forward inputs to contain `0`.  It might be fine to leave `backward` of `chainerx.prod` totally unimplemented.

Reply:
Sure. Do note that it'll take me some time to resume work on this PR.